### PR TITLE
Adds firebots to the robotics arsenal

### PR DESCRIPTION
**Pull request by Lett1**
## About The Pull Request

Adds a new type of bot: the Firebot!

This lil guy will scoot around and tries his best to extinguish burning people and hotspots around him.
If emagged he will knock people over and wets the floor under em for extra salt.

Sprites by Nemvar on discord. Sfx by generic windows tts.

## Why It's Good For The Game

Firebots are supposed to be used by mapmakers/people in highly flammable enviroments, like chemistry or toxins. The current way of fighting fires is to either drain the air out of the room or to go after it with extinguishers. The bot is supposed to support the later, he is not supposed to be a cure all for dangerous plasma fires.

## Changelog
:cl:
add: Added a new type of bot: the firebot.
add: Firebots can be constructed out of extinguisher + borg arm > red fire helmet > prox sensor
/:cl:

### CODE REVIEW

**Comment by davethwave** [comment]
Seems like a ok idea.

**Comment by habader** [comment]
Add a slime murdering option

**Comment by tralezab** [comment]
the emagged version just knocks people over? nothing to do with fire?

**Comment by Lett1** [comment]
> the emagged version just knocks people over? nothing to do with fire?

I think setting people on fire or making fire spread more would be to dangerous for something like that since fire can kill rather quickly.

**Comment by ShizCalev** [comment]
doesn't drop the fire helmet? :(

**Comment by anconfuzedrock** [comment]
why firesuit helmets? aren't those like rare as fuck? unless you mean just red helmets from emergency lockers in which case I swear to god if I start opening fire closets and all the helmets are gone I'm going to gut somebody

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+/mob/living/simple_animal/bot/firebot/explode()
+	on = FALSE
+	visible_message("<span class='boldannounce'>[src] blows apart!</span>")
+	var/atom/Tsec = drop_location()
+
+	//drop_part(toolbox, Tsec)
+
+	new /obj/item/assembly/prox_sensor(Tsec)
+
+	var/turf/T = get_turf(loc)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]: get_turf(src)

Comment (Lett1) [reply code/modules/mob/living/simple_animal/bot/firebot.dm]: Fixed, also drops the helmet now.

**Comment by Lett1** [comment]
> why firesuit helmets? aren't those like rare as fuck? unless you mean just red helmets from emergency lockers in which case I swear to god if I start opening fire closets and all the helmets are gone I'm going to gut somebody

Yeah I mean those, gonna edit the text.

**Comment by MrPerson** [comment]
>I swear to god if I start opening fire closets and all the helmets are gone I'm going to gut somebody

Peak SS13 gameplay right there. Maybe we should spawn some of the lockers missing stuff?

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+	else if(stationary_mode) //Bot has yellow light to indicate stationary mode.
+		icon_state = "firebots1"
+	else
+		icon_state = "firebot1"
+
+
+/mob/living/simple_animal/bot/firebot/explode()
+	on = FALSE
+	visible_message("<span class='boldannounce'>[src] blows apart!</span>")
+	var/atom/loc = drop_location()
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]: don't name this var loc. :(

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+	new /obj/item/assembly/prox_sensor(loc)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]: .

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+	new /obj/item/clothing/head/hardhat/red(loc)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]: .

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+	var/turf/T = get_turf(loc)
+	if(isopenturf(T))
+		var/turf/open/theturf = T
+		theturf.MakeSlippery(TURF_WET_WATER, min_wet_time = 10 SECONDS, wet_time_to_add = 5 SECONDS)
+
+	if(prob(50))
+		drop_part(robot_arm, loc)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]: no.

**Review by ShizCalev — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+
+/mob/living/simple_animal/bot/firebot/explode()
+	on = FALSE
+	visible_message("<span class='boldannounce'>[src] blows apart!</span>")
+	var/atom/loc = drop_location()
+
+	new /obj/item/assembly/prox_sensor(loc)
+	new /obj/item/clothing/head/hardhat/red(loc)
+
+	var/turf/T = get_turf(loc)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]: get_turf(src) still

**Comment by OverlyGeneric** [comment]
Goonstation's **Leaner** Cousin

**Comment by ATH1909** [comment]
Idea here: Instead of having the firebot knock you down directly and then dump water on the ground, it should dump water on the ground and then swap tiles with you (even though bots don't normally swap tiles with mobs, IIRC), slipping you like how a humanoid with a fire extinguisher would.

**Comment by OverlyGeneric** [comment]
> Idea here: Instead of having the firebot knock you down directly and then dump water on the ground, it should dump water on the ground and then swap tiles with you (even though bots don't/can't normally swap tiles with mobs, IIRC), slipping you like a humanoid with a fire extinguisher would.

I still think the thing should be able to wail onto the poor bastard that steps into the red bull's way. 

However, you could make this possible by having it move onto the same tile as the player like most other friendly mobs do when they randomly stumble around.

**Comment by GunHog** [comment]
Slipping people is the cleanbot's thing, though...

**Comment by PKPenguin321** [comment]
\>emagging it doesn't make it shoot fire
Lame!

**Comment by OverlyGeneric** [comment]
> >emagging it doesn't make it shoot fire
> >emagging doesn't make it shoot snow and freeze people to death as it beats everyone with its robust body.
> >lamer

**Comment by dr-bee** [comment]
Having a firebot that doesnt shoot fire somehow is a waste of a good pun. Also red fire helmets arent mass printable making a hard limit to the amount of firebots that can be built. Consider another item or make the helmets printable at an autolathe or order-able from cargo.

**Comment by Lett1** [comment]
> Having a firebot that doesnt shoot fire somehow is a waste of a good pun. Also red fire helmets arent mass printable making a hard limit to the amount of firebots that can be built. Consider another item or make the helmets printable at an autolathe or order-able from cargo.

You can order a firefighting crate from cargo, it contains two helmets.

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+	else
+		icon_state = "firebot1"
+
+
+/mob/living/simple_animal/bot/firebot/explode()
+	on = FALSE
+	visible_message("<span class='boldannounce'>[src] blows apart!</span>")
+
+	new /obj/item/assembly/prox_sensor(src.loc)
+	new /obj/item/clothing/head/hardhat/red(src.loc)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]: use drop_location(). You had this all right the first time with `Tsec`, not sure why you keep changing it. Refer to the below code.

https://github.com/tgstation/tgstation/blob/a9ca93b60dbf7b9f0fc5cbe94c406a2126312fee/code/modules/mob/living/simple_animal/bot/floorbot.dm#L361-L379

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+		icon_state = "firebots1"
+	else
+		icon_state = "firebot1"
+
+
+/mob/living/simple_animal/bot/firebot/explode()
+	on = FALSE
+	visible_message("<span class='boldannounce'>[src] blows apart!</span>")
+
+	new /obj/item/assembly/prox_sensor(src.loc)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]: .

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+	new /obj/item/clothing/head/hardhat/red(src.loc)
+
+	var/turf/T = get_turf(src)
+	if(isopenturf(T))
+		var/turf/open/theturf = T
+		theturf.MakeSlippery(TURF_WET_WATER, min_wet_time = 10 SECONDS, wet_time_to_add = 5 SECONDS)
+
+	if(prob(50))
+		drop_part(robot_arm, src.loc)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]: .

**Review by ShizCalev — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+		flick("firebot1_use", user)
+	internal_ext.afterattack(target, user, null)
+
+/mob/living/simple_animal/bot/firebot/update_icon()
+	if(!on)
+		icon_state = "firebot0"
+		return
+	if(IsStun() || IsParalyzed())
+		icon_state = "firebots1"
+		return
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]: this return isn't doing anything

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+/mob/living/simple_animal/bot/firebot/process_scan(atom/scan_target)
+	var/result
+
+	if(scan_target == src)
+		return result
+
+	if(is_burning(scan_target))
+		if((detected_cooldown + DETECTED_VOICE_INTERVAL) < world.time)
+			speak("Fire detected!")
+			playsound(src.loc, "sound/voice/firebot/detected.ogg", 50, 0)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]:
```suggestion
			playsound(src, "sound/voice/firebot/detected.ogg", 50)
```

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+				playsound(src.loc, "sound/voice/firebot/extinguishing.ogg", 50, 0)
+			speech_cooldown = world.time
+
+		if(emagged == 2) // When emagged, knock people down and wet the floors under em
+			var/mob/living/M = target_fire
+			M.Knockdown(3)
+			src.do_attack_animation(M)
+			visible_message("<span class='danger'>[src] rams into [M] knocking them over!</span>")
+
+			var/turf/T = get_turf(M.loc)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]:
```suggestion
			var/turf/T = get_turf(M)
```

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+			else
+				speak("Extinguishing!")
+				playsound(src.loc, "sound/voice/firebot/extinguishing.ogg", 50, 0)
+			speech_cooldown = world.time
+
+		if(emagged == 2) // When emagged, knock people down and wet the floors under em
+			var/mob/living/M = target_fire
+			M.Knockdown(3)
+			src.do_attack_animation(M)
+			visible_message("<span class='danger'>[src] rams into [M] knocking them over!</span>")
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]:
```suggestion
			visible_message("<span class='danger'>[src] rams into [M] knocking [M.p_them()] over!</span>")
```

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+
+	// Target reached ENGAGE WATER CANNON
+	if(target_fire && (get_dist(src, target_fire) <= (emagged == 2 ? 1 : 2))) // Make the bot spray water from afar when not emagged
+		if((speech_cooldown + SPEECH_INTERVAL) < world.time)
+			if(ishuman(target_fire))
+				speak("Stop, drop and roll!")
+				playsound(src.loc, "sound/voice/firebot/stopdropnroll.ogg", 50, 0)
+			else
+				speak("Extinguishing!")
+				playsound(src.loc, "sound/voice/firebot/extinguishing.ogg", 50, 0)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]:
```suggestion
				playsound(src, "sound/voice/firebot/extinguishing.ogg", 50)
```

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+			target_fire = scan(/turf/open, old_target_fire, scan_range) // Scan for burning turfs second
+
+		old_target_fire = target_fire
+
+	// Target reached ENGAGE WATER CANNON
+	if(target_fire && (get_dist(src, target_fire) <= (emagged == 2 ? 1 : 2))) // Make the bot spray water from afar when not emagged
+		if((speech_cooldown + SPEECH_INTERVAL) < world.time)
+			if(ishuman(target_fire))
+				speak("Stop, drop and roll!")
+				playsound(src.loc, "sound/voice/firebot/stopdropnroll.ogg", 50, 0)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]:
```suggestion
				playsound(src, "sound/voice/firebot/stopdropnroll.ogg", 50, 0)
```

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+		return
+
+	if(prob(1) && target_fire == null)
+		var/list/messagevoice = list("No fires detected." = 'sound/voice/firebot/nofires.ogg',
+		"Only you can prevent station fires." = 'sound/voice/firebot/onlyyou.ogg',
+		"Temperature nominal." = 'sound/voice/firebot/tempnominal.ogg',
+		"Keep it cool." = 'sound/voice/firebot/keepitcool.ogg')
+		var/message = pick(messagevoice)
+		speak(message)
+		playsound(loc, messagevoice[message], 50, 0)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]:
```suggestion
		playsound(src, messagevoice[message], 50)
```

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+
+	return dat
+
+/mob/living/simple_animal/bot/firebot/emag_act(mob/user)
+	..()
+	if(emagged == 2)
+		if(user)
+			to_chat(user, "<span class='danger'>[src] buzzes and beeps.</span>")
+		audible_message("<span class='danger'>[src] buzzes oddly!</span>")
+		playsound(src, "sparks", 75, 1)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]:
```suggestion
		playsound(src, "sparks", 75, TRUE)
```

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+		if(user)
+			old_target_fire = user
+		extinguish_fires = FALSE
+		extinguish_people = TRUE
+
+/mob/living/simple_animal/bot/firebot/Topic(href, href_list)
+	if(..())
+		return 1
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]:
```suggestion
		return TRUE
```

**Review by ShizCalev — CHANGES_REQUESTED** [review]

File: code/game/objects/items/extinguisher.dm [hunk 19]
```diff
@@ -229,3 +232,12 @@
...
 		user.visible_message("[user] empties out \the [src] onto the floor using the release valve.", "<span class='info'>You quietly empty out \the [src] using its release valve.</span>")
+
+/obj/item/extinguisher/attackby(obj/O, mob/user, params)
+	if(istype(O, /obj/item/bodypart/l_arm/robot) || istype(O, /obj/item/bodypart/r_arm/robot))
+		to_chat(user, "<span class='notice'>You add [O] to [src].</span>")
+		qdel(O)
+		qdel(src)
+		user.put_in_hands(new /obj/item/bot_assembly/firebot)
+	else
+		..()
```
Comment (ShizCalev) [inline code/game/objects/items/extinguisher.dm]: new line at the end.

**Comment by Lett1** [comment]
Sorry you have to put up with my shit @ShizCalev, lets hope it works this time.

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 20]
```diff
@@ -0,0 +1,308 @@
...
+		return
+
+	if(prob(1) && target_fire == null)
+		var/list/messagevoice = list("No fires detected." = 'sound/voice/firebot/nofires.ogg',
+		"Only you can prevent station fires." = 'sound/voice/firebot/onlyyou.ogg',
+		"Temperature nominal." = 'sound/voice/firebot/tempnominal.ogg',
+		"Keep it cool." = 'sound/voice/firebot/keepitcool.ogg')
+		var/message = pick(messagevoice)
+		speak(message)
+		playsound(loc, messagevoice[message], 50)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]:
```suggestion
 		playsound(loc, messagevoice[message], 50)
```

**Comment by ATH1909** [comment]
> Slipping people is the cleanbot's thing, though...

Oh, true.

Maybe have it shoot out frozen water vapor instead of the room temperature(?) water it normally shoots out, freezing large areas of the station?

File: code/modules/mob/living/simple_animal/bot/firebot.dm [hunk 21]
```diff
@@ -0,0 +1,308 @@
...
+				playsound(src, "sound/voice/firebot/stopdropnroll.ogg", 50, 0)
+			else
+				speak("Extinguishing!")
+				playsound(src, "sound/voice/firebot/extinguishing.ogg", 50, 0)
+			speech_cooldown = world.time
+
+		if(emagged == 2) // When emagged, knock people down and wet the floors under em
+			var/mob/living/M = target_fire
+			M.Knockdown(3)
+			src.do_attack_animation(M)
```
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/bot/firebot.dm]: remove src

**Comment by anconfuzedrock** [comment]
(((I WAS SERIOUS ABOUT THE FIREFIGHTER HELMETS A CLEANBOT ONLY NEEDS A BUCKET A MEDIBOT ONLY NEEDS A MEDKIT FOR THE LOVE OF GOD AND ALL THAT IS HOLY DO NOT MAKE IT TAKE FIRE HELMETS THE SPRITE DOESN'T EVEN LOOK GOOD)))

**Comment by Lett1** [comment]
Changed how the emagged one works: it now sprays fire (cfl3) onto people.

File: code/game/objects/items/extinguisher.dm [hunk 22]
```diff
@@ -229,3 +232,12 @@
...
```
Comment (TerraGS) [inline code/game/objects/items/extinguisher.dm]: This part should be in construction.dm to keep it consistent with medbots and floorbots.

Comment (Lett1) [reply code/game/objects/items/extinguisher.dm]: Didn't know that you could craft em like this, added now.

**Comment by Lett1** [comment]
> Also needs a recipe in /code/modules/crafting/recipes.dm like the other bots.

Didn't know you could craft em like this, added it.

**Comment by DapperMola** [comment]
should have emagged lines like the cleanbot
"Burn, baby burn."
"Is it getting hot in here? Or is it just you?"
"Smoking kills."

**Pull request merged**